cmd/kubeadm/app/cmd/init.go

```diff
@@ -386,12 +386,33 @@ func newInitData(cmd *cobra.Command, args []string, options *initOptions, out io
 		// Validate that also the required kubeconfig files exists and are invalid, because
 		// kubeadm can't regenerate them without the CA Key
 		kubeconfigDir := options.kubeconfigDir
-		if options.dryRun {
-			kubeconfigDir = dryRunDir
-		}
 		if err := kubeconfigphase.ValidateKubeconfigsForExternalCA(kubeconfigDir, cfg); err != nil {
 			return nil, err
 		}
+
+		// If using an externalCA while dryrun, need to copy those files to dryrun dir for later use
+		if options.dryRun {
+			externalCAFiles := []string{
+				filepath.Join(cfg.CertificatesDir, kubeadmconstants.CACertName),
+				filepath.Join(kubeconfigDir, kubeadmconstants.AdminKubeConfigFileName),
+				filepath.Join(kubeconfigDir, kubeadmconstants.KubeletKubeConfigFileName),
+				filepath.Join(kubeconfigDir, kubeadmconstants.ControllerManagerKubeConfigFileName),
+				filepath.Join(kubeconfigDir, kubeadmconstants.SchedulerKubeConfigFileName),
+			}
+
+			for _, externalCAFile := range externalCAFiles {
+				data, err := os.ReadFile(externalCAFile)
+				if err != nil {
+					return nil, err
+				}
+				_, fileName := filepath.Split(externalCAFile)
+				err = os.WriteFile(filepath.Join(dryRunDir, fileName), data, os.FileMode(0600))
```

Comment:
we should preserve the FileMode of the original file that we read.

Reply:
we should, done.